pkg/networkservice/common/heal/client.go

```diff
@@ -66,69 +70,97 @@ func NewClient(client networkservice.MonitorConnectionClient, onHeal *networkser
 		updateExecutor:    serialize.NewExecutor(),
 		eventReceiver:     nil, // This is intentionally nil
 		recvEventExecutor: serialize.NewExecutor(),
 	}
+
 	if rv.onHeal == nil {
 		rv.onHeal = addressof.NetworkServiceClient(rv)
 	}
-	rv.updateExecutor.AsyncExec(func() {
-		runtime.SetFinalizer(rv, func(f *healClient) {
-			f.updateExecutor.AsyncExec(func() {
-				if f.cancelFunc != nil {
-					f.cancelFunc()
-				}
-			})
-		})
-		ctx, cancelFunc := context.WithCancel(context.Background())
-		rv.cancelFunc = cancelFunc
-		// TODO decide what to do about err here
-		recv, _ := rv.client.MonitorConnections(ctx, nil)
-		rv.eventReceiver = recv
-		rv.recvEventExecutor.AsyncExec(rv.recvEvent)
-	})
 
 	return rv
 }
 
+func (f *healClient) init() {
+	if f.eventReceiver != nil {
+		select {
+		case <-f.eventReceiver.Context().Done():
+			return
+		default:
+			f.eventReceiver = nil
+		}
+	}
+
+	logrus.Info("Creating new eventReceiver")
+
+	recv, err := f.client.MonitorConnections(f.chainContext, &networkservice.MonitorScopeSelector{}, grpc.WaitForReady(true))
+	if err != nil {
+		f.updateExecutor.AsyncExec(func() {
+			f.init()
+		})
+		return
+	}
+
+	f.eventReceiver = recv
+	f.recvEventExecutor.AsyncExec(f.recvEvent)
+}
+
+func (f *healClient) lazyInit(ctx context.Context, id string) {
+	f.updateExecutor.AsyncExec(func() {
+		if len(f.connectionIDs) == 0 {
+			f.chainContext = chaincontext.ChainContext(ctx)
+			f.init()
+		}
+		f.connectionIDs[id] = true
+	})
+}
+
 func (f *healClient) recvEvent() {
 	select {
 	case <-f.eventReceiver.Context().Done():
-		f.eventReceiver = nil
+		return
 	default:
 		event, err := f.eventReceiver.Recv()
-		if err != nil {
-			event = nil
-		}
 		f.updateExecutor.AsyncExec(func() {
+			if err != nil {
+				for id := range f.reported {
+					f.requestors[id]()
+					delete(f.reported, id)
+				}
+				f.init()
+				return
+			}
+
 			switch event.GetType() {
 			case networkservice.ConnectionEventType_INITIAL_STATE_TRANSFER:
 				f.reported = event.GetConnections()
+				if event.GetConnections() == nil {
+					f.reported = map[string]*networkservice.Connection{}
+				}
+
 			case networkservice.ConnectionEventType_UPDATE:
 				for _, conn := range event.GetConnections() {
 					f.reported[conn.GetId()] = conn
 				}
+
 			case networkservice.ConnectionEventType_DELETE:
-				for _, conn := range event.GetConnections() {
+				for id, conn := range event.GetConnections() {
+					f.requestors[id]()
 					delete(f.reported, conn.GetId())
-					if f.requestors[conn.GetId()] != nil {
-						f.requestors[conn.GetId()]()
-					}
-				}
-			}
-			for id, request := range f.requestors {
-				if _, ok := f.reported[id]; !ok {
-					request()
 				}
 			}
 		})
 	}
+
 	f.recvEventExecutor.AsyncExec(f.recvEvent)
 }
 
 func (f *healClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	f.lazyInit(ctx, request.GetConnection().GetId())
```

Comment:
It still looks like we are reIniting on each Request, why not AsyncExec to init in NewClient?

Reply:
We do it only for the first Request. But laziness is not crucial anymore since new per-chain Context implemented

Reply:
:)

Reply:
Fixed, get rid of laziness